File: 1. Create and print Binary Tree.cpp
/*
 Task: Create and print Binary Tree
 Time Complexity: O(n) — build and print visit each node once
 Space Complexity: O(n) — queue for level-order construction/printing
*/
#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    Node* left;
    Node* right;

    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};

Node* buildLevelOrder(const vector<int>& vals)
{
    if (vals.empty())
    {
        return nullptr;
    }

    int i = 0;
    Node* root = new Node(vals[i++]);
    queue<Node*> q;
    q.push(root);

    while (i < static_cast<int>(vals.size()))
    {
        Node* cur = q.front();
        q.pop();

        if (vals[i] != INT_MIN)
        {
            cur->left = new Node(vals[i]);
            q.push(cur->left);
        }

        ++i;
        if (i >= static_cast<int>(vals.size()))
        {
            break;
        }

        if (vals[i] != INT_MIN)
        {
            cur->right = new Node(vals[i]);
            q.push(cur->right);
        }

        ++i;
    }

    return root;
}

void printLevelOrder(Node* root)
{
    if (!root)
    {
        return;
    }

    queue<Node*> q;
    q.push(root);

    while (!q.empty())
    {
        Node* cur = q.front();
        q.pop();

        cout << cur->data << " ";
        if (cur->left)
        {
            q.push(cur->left);
        }
        if (cur->right)
        {
            q.push(cur->right);
        }
    }

    cout << endl;
}

int main()
{
    // use INT_MIN to represent nulls in level-order array
    vector<int> vals = {1, 2, 3, 4, 5, INT_MIN, 7};
    Node* root = buildLevelOrder(vals);
    cout << "Level order print: ";
    printLevelOrder(root);
    return 0;
}

/*
 Task: Create and print Binary Tree
 Time Complexity: O(n)
 Space Complexity: O(n)
*/
// sc tc

File: 1. Create and print Binary Tree.py
"""
Task: Create and print Binary Tree
Time Complexity: O(n)
Space Complexity: O(n)
"""

from collections import deque

class Node:
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None


def build_level_order(vals):
    if not vals: return None
    it=iter(vals)
    root=Node(next(it))
    q=deque([root])
    for v in it:
        cur=q[0]
        if v is not None:
            node=Node(v); cur.left=node; q.append(node)
        else:
            cur.left=None
        try:
            v2=next(it)
        except StopIteration:
            break
        if v2 is not None:
            node=Node(v2); cur.right=node; q.append(node)
        else:
            cur.right=None
        q.popleft()
    return root


def print_level_order(root):
    if not root: return
    q=deque([root])
    while q:
        cur=q.popleft()
        print(cur.data, end=' ')
        if cur.left: q.append(cur.left)
        if cur.right: q.append(cur.right)
    print()

if __name__=='__main__':
    vals=[1,2,3,4,5,None,7]
    root=build_level_order(vals)
    print('Level order print:', end=' ')
    print_level_order(root)

# Task: Create and print Binary Tree
# Time Complexity: O(n)
# Space Complexity: O(n)
